### PR TITLE
Review 3rd: Childhood Immunizations HEDIS Measure

### PR DESCRIPTION
**Pull request by TylerNoblett**
# Related Tickets

- [SAR-76](https://jira.amida-tech.com/browse/SAR-76)

# How Things Worked (or Didn't) Before This PR
There was no CQL and associated files for Childhood Immunizations

# How Things Work Now (And How to Test)
## How it works now
I've added the CQL and associated files for Childhood Immunizations
Additionally, since the only differences between the `exec` files are the imports, I've created an exec file template, added the functionality to output the results to a JSONL file, and added some tests

## To test
run `yarn` to install the necessary packages
run `node exec-files/exec-childhood-immunization-status.js`
### NOTE:
You'll get about 1000 lines of console error messages - you can ignore them for now. The console error isn't applicable due to the way our CQL is set up and is in an npm package. Because it's in an npm package, there's not an easy way to disable that error message that I was able to make it work at least) unless you do it locally. In production we won't have console logging available so it shouldn't be an issue

Use your favorite diffchecker to confirm your output matches what is below
```javascript
{
  'numer-EXM165-FHIR3': {
    Patient: FHIRObject {
      identifier: [Getter],
      active: [Getter],
      name: [Getter],
      telecom: [Getter],
      gender: [Getter],
      birthDate: [Getter],
      deceased: [Getter],
      address: [Getter],
      maritalStatus: [Getter],
      multipleBirth: [Getter],
      photo: [Getter],
      contact: [Getter],
      communication: [Getter],
      generalPractitioner: [Getter],
      managingOrganization: [Getter],
      link: [Getter],
      text: [Getter],
      contained: [Getter],
      extension: [Getter],
      modifierExtension: [Getter],
      id: [Getter],
      meta: [Getter],
      implicitRules: [Getter],
      language: [Getter]
    },
    'Qualifying Encounter': [ [FHIRObject] ],
    'Initial Population': true,
    'DTaP Immunizations or Procedures': [
      [FHIRObject],
      [FHIRObject],
      [FHIRObject],
      [FHIRObject],
      [FHIRObject],
      [FHIRObject],
      [FHIRObject],
      [FHIRObject]
    ],
    'DTaP Vaccinations': [ [FHIRObject], [FHIRObject] ],
    'Numerator Compliant DTaP Conditions': [],
    'DTaP Criteria Met': true,
    'Hepatitis A Vaccinations': [ [FHIRObject] ],
    'Numerator Compliant Hepatitis A Conditions': [ [FHIRObject], [FHIRObject] ],
    'Hepatitis A Criteria Met': true,
    'Hepatitis B Immunizations or Procedures': [ [FHIRObject], [FHIRObject], [FHIRObject], [FHIRObject] ],
    'Hepatitis B Vaccinations': [ [FHIRObject], [FHIRObject] ],
    'Numerator Compliant Hepatitis B Conditions': [ [FHIRObject] ],
    'Hepatitis B Criteria Met': true,
    'HIB Immunizations or Procedures': [ [FHIRObject], [FHIRObject], [FHIRObject], [FHIRObject] ],
    'HIB Vaccinations': [ [FHIRObject], [FHIRObject] ],
    'Numerator Compliant HIB Conditions': [ [FHIRObject] ],
    'HIB Criteria Met': true,
    'Influenza Immunizations or Procedures': [ [FHIRObject], [FHIRObject], [FHIRObject] ],
    'Influenza Vaccinations': [ [FHIRObject], [FHIRObject] ],
    'Numerator Compliant Influenza Conditions': [ [FHIRObject] ],
    'Influenza Criteria Met': true,
    'IPV Immunizations or Procedures': [ [FHIRObject], [FHIRObject], [FHIRObject], [FHIRObject] ],
    'IPV Vaccinations': [ [FHIRObject], [FHIRObject] ],
    'Numerator Compliant IPV Conditions': [ [FHIRObject] ],
    'IPV Criteria Met': true,
    'Measles Indicators': [ [FHIRObject] ],
    'MMR Vaccinations': [ [FHIRObject] ],
    'Mumps Indicators': [ [FHIRObject] ],
    'Numerator Compliant MMR Conditions': [ [FHIRObject] ],
    'Rubella Indicators': [ [FHIRObject] ],
    'MMR Criteria Met': true,
    'Pneumococcal Conjugate Immunizations or Procedures': [ [FHIRObject], [FHIRObject], [FHIRObject], [FHIRObject] ],
    'Pneumococcal Conjugate Vaccinations': [ [FHIRObject] ],
    'Numerator Compliant Pneumococcal Conjugate Conditions': [ [FHIRObject] ],
    'Pneumococcal Conjugate Criteria Met': true,
    'Rotavirus 2 Dose Immunizations or Procedures': [ [FHIRObject], [FHIRObject] ],
    'Rotavirus 2 Dose Vaccinations': [ [FHIRObject] ],
    'Rotavirus 3 Dose Immunizations or Procedures': [ [FHIRObject], [FHIRObject], [FHIRObject] ],
    'Rotavirus 2 or 3 Dose Vaccinations A': [ [FHIRObject], [FHIRObject] ],
    'Rotavirus 2 or 3 Dose Vaccinations B': [],
    'Rotavirus 2 or 3 Dose Vaccinations C': [],
    'Rotavirus 2 or 3 Dose Vaccinations': [ [FHIRObject], [FHIRObject] ],
    'Rotavirus 3 Dose Vaccinations': [ [FHIRObject] ],
    'Numerator Compliant Rotavirus Conditions': [ [FHIRObject] ],
    'Rotavirus Criteria Met': true,
    'VZV Vaccinations': [ [FHIRObject] ],
    'Numerator Compliant VZV Conditions': [ [FHIRObject], [FHIRObject] ],
    'VZV Criteria Met': true,
    Numerator: 1,
    'Denominator Exclusions': false,
    Denominator: 1
  }
}
{
  'Denominator Count': 1,
  'Numerator Count': 1,
  'CDC Measure Score': 1
} 
```
## To Review
`exec-cis.js` - This is the exec file that actually does the work - there's not much to review here.

`cis-codese.js` - The list of codes in each value set. This is programmatically generated using the CQL from the [VSAC](https://vsac.nlm.nih.gov/) API, so there's nothing to review here

`cis.json` - JSON version of the CQL created using the [clinical_quality_language](https://github.com/cqframework/clinical_quality_language) - nothing to review here

`cis-patients.js` - This contains the fake patients and their data. If you are really gung-ho, you can review the output of the results and confirm that the expressions with the CQL library returned the correct response based on the data, but that might take a while.

`cis.cql` - this is the main piece of the PR. Read through the following requirements, and as best you can, confirm that the CQL matches the guidelines. If you'd prefer, whenever you get to this portion I can send you the word doc these requirements are in, which might be easier than looking at screenshots.
Review the `.cql` file based on the HEDIS measures
Screen Shot 2021-01-19 at 4 29 49 PM
Screen Shot 2021-01-19 at 4 29 59 PM
Screen Shot 2021-01-19 at 4 30 10 PM
Screen Shot 2021-01-19 at 4 30 19 PM
Screen Shot 2021-01-19 at 4 30 29 PM
Screen Shot 2021-01-19 at 4 30 39 PM
Screen Shot 2021-01-19 at 4 30 54 PM

### CODE REVIEW

**Review by eawest2 — APPROVED** [review]
The code works, and the tests perform as expected with the expected output, so I'm approving.

However:

I do get several hundred, or possibly a few thousand (I can't scroll back far enough to find the end) of these errors in the screenshot. I don't know if this is an artifact of the tests, but whatever it's doing we probably don't need to do without a conditional activation if it's expected to error like that nearly every time.

image

**Comment by TylerNoblett** [comment]
> The code works, and the tests perform as expected with the expected output, so I'm approving.
> 
> However:
> 
> I do get several hundred, or possibly a few thousand (I can't scroll back far enough to find the end) of these errors in the screenshot. I don't know if this is an artifact of the tests, but whatever it's doing we probably don't need to do without a conditional activation if it's expected to error like that nearly every time.
> 
> image

Ah good catch - I need to put a note about that. I've disabled that error locally, but it's an error message we're getting from an npm package, so it's not something that we can remove in a scalable way. We'll also disable logging in production, but I totally forgot that everyone else would get that error!

File: exec-files/exec-template.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  const executor = new cql.Executor(lib, cservice);
+  const patientSource = cqlfhir.PatientSource.FHIRv400();
+  patientSource.loadBundles(patients);
+
+  const result = executor.exec(patientSource);
+  console.log(result.patientResults); // eslint-disable-line no-console
+  console.log(result.unfilteredResults); // eslint-disable-line no-console
+
+  fs.writeFile(
```
Comment (TylerNoblett) [inline exec-files/exec-template.js:23]: James needs the output to be a json file, so I added that functionality. Additionally, since we'll be grabbing the result files from thousands of nodes, I added a UUID to the filenames to ensure they're unique

File: exec-files/exec-cdc_diabetes-bp.js [hunk 2]
```diff
@@ -1,28 +1,6 @@
...
-  )
-};
-
-const executor = new cql.Executor(lib, cservice, parameters);
-const patientSource = cqlfhir.PatientSource.FHIRv400();
-patientSource.loadBundles(patients);
-
-const result = executor.exec(patientSource);
-console.log(result); // eslint-disable-line no-console
+execute(measure, patients, codeservice);
```
Comment (TylerNoblett) [inline exec-files/exec-cdc_diabetes-bp.js:6]: all of the changes to the exec files should be in another PR, but since this PR was open I started adding small changes and here we are 😬🤷‍♂️

File: cql/childhood-immunization-status.cql [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+		"DTaP Immunizations or Procedures" Vaccination4
+		where Coalesce(Vaccination2.performed.start, Vaccination2.occurrence) 1 day or more after Coalesce(Vaccination1.performed.start, Vaccination1.occurrence)
+			and Coalesce(Vaccination3.performed.start, Vaccination3.occurrence) 1 day or more after Coalesce(Vaccination2.performed.start, Vaccination2.occurrence) 
+			and Coalesce(Vaccination4.performed.start, Vaccination4.occurrence) 1 day or more after Coalesce(Vaccination3.performed.start, Vaccination3.occurrence)
+		return Vaccination1
+
+define "Numerator Compliant DTaP Conditions":
+	( ["Condition": "Anaphylactic Reaction to DTaP Vaccine"]
+		union ["Condition": "Encephalopathy due to Childhood Vaccination"] ) DPaTConditions
+		where AgeInDaysAt(Coalesce(DPaTConditions.onset.start, DPaTConditions.onset))<= 73
```
Comment (JamesbZhu) [inline cql/childhood-immunization-status.cql:121]: this seems like it belongs in the exclusion criteria, and i don't see a condition in the spec that specifies AgeInDaysAt()< 73

File: cql/childhood-immunization-status.cql [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+	return HepA
+
+define "Numerator Compliant Hepatitis A Conditions":
+	( ( ["Observation": "Anti Hepatitis A IgG Antigen Test"] HepAAntigenTest
+			with HepAAntigenTest.value.coding coding such that exists(
+							HepAAntigenTest.value.coding coding
+              	where coding.code in "Positive Finding"
+        )
+	)
+		union ( ["Condition": "Anaphylactic Reaction to Hepatitis A Vaccine"]
```
Comment (JamesbZhu) [inline cql/childhood-immunization-status.cql:143]: Anaphylactic reaction seems like it should be in the exclusion criteria too

File: cql/childhood-immunization-status.cql [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+	["Condition": "Anaphylactic Reaction to Hemophilus Influenza B (HiB) Vaccine"] HIBReaction
+		where AgeInDaysAt(Coalesce(HIBReaction.onset.start, HIBReaction.onset))<= 730
+
+define "HIB Criteria Met":
+	exists("HIB Vaccinations") or exists("Numerator Compliant HIB Conditions")
+
+define "Influenza Immunizations or Procedures":
+	( ["Immunization": "Influenza Vaccine"]
+		union ["Procedure": "Influenza Vaccine Administered"] ) FluVaccination
+		where AgeInDaysAt(Coalesce(FluVaccination.performed.start, FluVaccination.occurrence))>= 42
```
Comment (JamesbZhu) [inline cql/childhood-immunization-status.cql:209]: it looks like this should be >= 180

File: cql/childhood-immunization-status.cql [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+		and "DTaP Criteria Met"
+		and "IPV Criteria Met"
+		and "MMR Criteria Met"
+		and "HIB Criteria Met"
+		and "Hepatitis B Criteria Met"
+		and "VZV Criteria Met"
+		and "Pneumococcal Conjugate Criteria Met"
+		and "Hepatitis A Criteria Met"
+		and "Rotavirus Criteria Met"
+		and "Influenza Criteria Met"
```
Comment (JamesbZhu) [inline cql/childhood-immunization-status.cql:470]: This seems like it would cover combination 10 only, though the verbage in the HEDIS specifcation isn't very clear on how combinations 2 through combination 10 are calculated w/ respect to the CIS measure score as a whole

File: exec-files/exec-template.js (hunk 1 above)
Comment (JamesbZhu) [reply exec-files/exec-template.js:61]: awesome!!

File: cql/childhood-immunization-status.cql (hunk 3 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:121]: ahhh - I had been ignoring the exclusion section because it was optional, but now that I read through I see that this is an exclusion. oops - that should be 730, which is two years in days

File: cql/childhood-immunization-status.cql (hunk 4 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:143]: nice! Yeah I've been moving all of the reaction stuff for each category into an exclusion expression and hope to finish that soon

File: cql/childhood-immunization-status.cql (hunk 5 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:209]: good catch - changed it!

File: cql/childhood-immunization-status.cql (hunk 6 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:470]: yeah - If any of the combos are valid, then it seems that I (from a logical perspective) don't even care about PCV, Hep A, RV, or the flu, if you only need the prior 6 vaccines to 'pass' - I've commented out the last 4, but I think it will help to get input from Mike or a SME

File: cql/childhood-immunization-status.cql (hunk 3 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql]: changed it to be in the exclusion criteria!

File: cql/childhood-immunization-status.cql (hunk 4 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql]: moved it into the exclusion criteria!

File: cql/childhood-immunization-status.cql [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+		and "DTaP Criteria Met"
+		and "IPV Criteria Met"
+		and "MMR Criteria Met"
+		and "HIB Criteria Met"
+		and "Hepatitis B Criteria Met"
+		and "VZV Criteria Met"
+		//and "Pneumococcal Conjugate Criteria Met"
+		//and "Hepatitis A Criteria Met"
+		//and "Rotavirus Criteria Met"
+		//and "Influenza Criteria Met"
```
Comment (TylerNoblett) [inline cql/childhood-immunization-status.cql:436]: if you look at the logic in the HEDIS spec, and assuming that all of the 9 combinations are equally valid, it seems that we don't even need to check to these last 4 vaccines - does that seem correct to you guys?

Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:436]: Additionally, there is this super confusing exclusion sentence that I'm not sure how to read:
>Exclude children who had a contraindication for a specific vaccine from the denominator for all antigen rates and the combination rates. The denominator for all rates must be the same.

File: cql/childhood-immunization-status.cql [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+valueset "Newborn Hepatitis B Vaccine Administered":'2.16.840.1.113883.3.464.1004.1397'
+valueset "Anaphylactic Reaction to Common Baker's Yeast": '2.16.840.1.113883.3.464.1003.199.12.1032'
+*/
+
+parameter "Measurement Period" default Interval[@2018-12-31T00:00:00.0, @2019-12-31T00:00:00.0)
+
+context Patient
+
+define "Initial Population":
+	exists ( "Qualifying Encounter" )
```
Comment (JamesbZhu) [inline cql/childhood-immunization-status.cql:96]: Can you point out where in the specification this exists? i can't seem to find it

File: cql/childhood-immunization-status.cql (hunk 7 above)
Comment (JamesbZhu) [reply cql/childhood-immunization-status.cql:436]: hm the verbage is a bit confusing, i'm not sure if they mean calculate each combination separately or if they mean each combination is equally valid - i think a third opinion would be nice here

**Review by JamesbZhu — COMMENTED** [review]
Only part i had a question on was the Initial population query. The calculation logic for combinations 2-10 also seems unclear to me, it'll be valuable to have a third opinion or SME validation (when we get to that point). Otherwise it looks great!

File: cql/childhood-immunization-status.cql (hunk 8 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:96]: ah - this was in the original CQL I found online - looking through the HEDIS spec I can't find it - good catch!

File: exec-files/exec-template.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+var { cloneDeep } = require('lodash');
+const { v4: uuidv4 } = require('uuid');
+const codes = require('../src/cql-code-service');
+const cql = require('../src/cql');
+const fhirhelpers = require('../json-elm/FHIRHelpers.json');
+
+const removeArrayValues = patient => {
+  const clonedPatient = cloneDeep(patient);
+  Object.entries(clonedPatient).forEach(([propertyKey, propertyValue]) => {
+    // remove property values that are arrays - the data pipeline doesn't need them
```
Comment (TylerNoblett) [inline exec-files/exec-template.js:13]: @JamesbZhu here I'm removing any property values that are lists (since you only calculate things based off of bools). Does me removing any attributes that are lists seem like it could be a bad idea at all? It seems that if someone is writing CQL, we'd just need to instruct them to make as many expressions as possible to check to see if a value exists, rather than returning the value - I guess I'm just nervous that we might lose an important metric if we forget to write a particular expression to return a boolean - does that seem like an issue to you? IDK - we can talk later if it's helpful!

Comment (TylerNoblett) [reply exec-files/exec-template.js:13]: also, I'm realizing that 1) maybe we should go back through each CQL file and make sure that each expression name is as clear as possible now that those expressions are going to show up in the dashboard / will the customer know what those expression names mean (probably a question for MIke)? IF they've read the HEDIS spec for a particular measure then they will, but if they haven't I wonder if that information will be helpful - maybe something to talk through? IDK

Comment (TylerNoblett) [reply exec-files/exec-template.js:13]: @JamesbZhu @hmayenve - do you guys have any thoughts about the two comments I posted above?

File: cql/childhood-immunization-status.cql [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+		//and "Hepatitis A Criteria Met"
+		//and "Rotavirus Criteria Met"
+		//and "Influenza Criteria Met"
+	)
+		when true then 1
+    else 0
+  end
+
+define "Denominator":
+  case ("Initial Population" and not "Denominator Exclusions")
```
Comment (hmayenve) [inline cql/childhood-immunization-status.cql:438]: For denominator, the spec says "The eligible population." does the "Denominator Exclusions" come from the "Exclusion (optional)" section of the spec? im not sure where this is coming from.

File: cql/childhood-immunization-status.cql [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+		union [Procedure: "DTaP Vaccine Administered"]) DTaPVaccination
+			where AgeInDaysAt(Coalesce(DTaPVaccination.performed.start, DTaPVaccination.occurrence))>= 42
+			and AgeInDaysAt(Coalesce(DTaPVaccination.performed.start, DTaPVaccination.occurrence))<= 730
+
+define "DTaP Vaccinations":
+	from
+		"DTaP Immunizations or Procedures" Vaccination1,
+		"DTaP Immunizations or Procedures" Vaccination2,
+		"DTaP Immunizations or Procedures" Vaccination3,
+		"DTaP Immunizations or Procedures" Vaccination4
```
Comment (hmayenve) [inline cql/childhood-immunization-status.cql:106]: The spec says "At least four DtaP vaccinations ....", that's why we do this, got it.

File: cql/childhood-immunization-status.cql [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+	( [Immunization: "Hepatitis B Vaccine"]
+		union [Procedure: "Hepatitis B Vaccine Administered"]) HepBVaccination
+		where (AgeInDaysAt(Coalesce(HepBVaccination.performed.start, HepBVaccination.occurrence))<= 730
+		or AgeInDaysAt(Coalesce(HepBVaccination.performed.start, HepBVaccination.occurrence)) < 8)
+
+define "Hepatitis B Vaccinations":
+	from
+		"Hepatitis B Immunizations or Procedures" Vaccination1,
+		"Hepatitis B Immunizations or Procedures" Vaccination2,
+		"Hepatitis B Immunizations or Procedures" Vaccination3
```
Comment (hmayenve) [inline cql/childhood-immunization-status.cql:150]: The spec says "At least three HiB vaccinations ....", that's why we do this, got it.

File: cql/childhood-immunization-status.cql (hunk 7 above)
Comment (hmayenve) [reply cql/childhood-immunization-status.cql:431]: @TylerNoblett can you expand on the logic in the HEDIS specs? Im not fully grasping this part of the spec/cql.

File: cql/childhood-immunization-status.cql (hunk 10 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: ah yes! if you go down further there is a section entitled "Exclusion (optional)" that includes these things

File: cql/childhood-immunization-status.cql (hunk 7 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:431]: @hmayenve haha I will try, but I'm not really sure myself - I think this chart shows the multiple ways someone can achieve the "immunized" status - however, from a logical perspective, it seems that since combination 2 is included in every other combination, all you really need to test for is what is included in combination 2. Does that make sense?

File: cql/childhood-immunization-status.cql (hunk 10 above)
Comment (hmayenve) [reply cql/childhood-immunization-status.cql:438]: Are you referring to this one? if yes, im not see how this ties up with the expression.
Screen Shot 2021-02-23 at 9 49 46 AM

File: cql/childhood-immunization-status.cql (hunk 7 above)
Comment (hmayenve) [reply cql/childhood-immunization-status.cql:431]: ahh, yeah it makes more sense now. Thank you. Given that, I agree with you that we do not need to check for the last 4 vaccines.

File: cql/childhood-immunization-status.cql (hunk 10 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: ahh great question! The hospice stuff is mentioned here
Screen Shot 2021-02-23 at 10 09 01 AM
I'm not really sure what the top two bullets mean:
- contraindication:
  - In medicine, a contraindication is a condition that serves as a reason to not take a certain medical treatment due to the harm that it would cause the patient. Contraindication is the opposite of indication, which is a reason to use a certain treatment.
- I'm not really sure what a contraindication would look like for immunization
- Both sentences are generally confusing - I'm thinking we might need a SME to help clarify?


and all of the vaccine specific issues are addressed like so:
```
define "IPV Criteria Met":
	exists("IPV Vaccinations") and not exists("IPV Exclusions")
```

Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: Is this explanation helpful?

Comment (hmayenve) [reply cql/childhood-immunization-status.cql:438]: ahhh the expression is just checking for "Hospice care" in the patient's record and if found we exclude it from the eligible population and therefore the denominator. got it.

Comment (hmayenve) [reply cql/childhood-immunization-status.cql:438]: yeah it helped ty.

Comment (hmayenve) [reply cql/childhood-immunization-status.cql:438]: is the  "Exclusion (optional)" section something we want to address in this pr?

Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: Good point! teh CQL I found online had a most of the exclusion criteria in it, so I just left it in, but we can remove it if you think that would be better! As long as we save it I don't care too much and am not sure if it matters (at least right now)

Comment (hmayenve) [reply cql/childhood-immunization-status.cql:438]: the "Denominator Exclusions" expression addresses the "Note"
Screen Shot 2021-02-23 at 10 48 11 AM

you're saying that the CQL found online had some of the "Exclusion (optional)" criteria and that it's included in this PR?
Screen Shot 2021-02-23 at 9 49 46 AM

Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: the CQL I found online included the exclusion criteria for all of the vaccines - ie "IPV Exclusions"
```
define "IPV Criteria Met":
	exists("IPV Vaccinations") and not exists("IPV Exclusions")
```

Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: and I don't see what you're referring to - would you mind explaining in more detail?
> the "Denominator Exclusions" expression addresses the "Note"

generally speaking though, I don't think it would be a bad idea (since we're unsure of some of the exclusion criteria) if we just commented out the vaccine specific exclusions and came back to them at some point in the future. @JamesbZhu what do you do think?

Comment (hmayenve) [reply cql/childhood-immunization-status.cql:438]: sure. so i agree that the "Denominator Exclusions" expression addresses the "Note". Does the "IPV Exclusions" address the "Exclusion (optional)"?

**Comment by hmayenve** [comment]
Should we also add yarn.lock and package.json to the .gitignore file @TylerNoblett

File: cql/childhood-immunization-status.cql (hunk 10 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: yeah - all of the vaccines have a "{Vaccine name} Exclusions" expression which are referenced in the "Exclusion (optional)" section

**Comment by TylerNoblett** [comment]
> Should we also add yarn.lock and package.json to the .gitignore file @TylerNoblett

I don't feel that we need to, personally - I looked at 5 of our other repos and only one of them had the 'lock' file ignored so going with convention I don't think we need to.

Also, a JS repo always includes package.json (that is how you know what packages to install etc). If you have a reason for wanting to ignore them, I'd be happy to hear it though!

**Comment by hmayenve** [comment]
> > Should we also add yarn.lock and package.json to the .gitignore file @TylerNoblett
> 
> I don't feel that we need to, personally - I looked at 5 of our other repos and only one of them had the 'lock' file ignored so going with convention I don't think we need to.
> 
> Also, a JS repo always includes package.json (that is how you know what packages to install etc). If you have a reason for wanting to ignore them, I'd be happy to hear it though!

Got it.  No reason. Im okay with leaving them then.

File: cql/childhood-immunization-status.cql (hunk 10 above)
Comment (TylerNoblett) [reply cql/childhood-immunization-status.cql:438]: ahh - IPV Exclusions does address the IPV exclusion criteria in "Exclusion (optional)"

File: exec-files/exec-template.js (hunk 9 above)
Comment (JamesbZhu) [reply exec-files/exec-template.js:13]: Hey, just saw this - atleast for CIS i think we'd be ok leaving out the lists, the important metrics that we'd want to have remain as a bool (for the reports atleast) are the expressions that are included in inclusion of "Numerator" and "Denominator". as long as those are clear, at least for a MVP version of Saraswati dashboards i think it'll be completely ok. Not sure if that will change in the future but for the time being im ok with this.

Comment (JamesbZhu) [reply exec-files/exec-template.js:13]: to your second point, renaming the CQL expressions to something that is more clear would be ideal! it would make the dashboards a lot easier to read

File: cql/childhood-immunization-status.cql (hunk 10 above)
Comment (JamesbZhu) [reply cql/childhood-immunization-status.cql:438]: yup, at least in the near term i think it'll be ok to comment those exclusions out - but we should note that somewhere in the doc for when have an SME resource to get some validation

**Review by JamesbZhu — APPROVED** [review]
LGTM - great job! I think what's left is just a couple of small decisions that we'd need to make (e.g. if we're going to go with commenting out the exclusions for now).  

I'm still a little unclear as to how the combinations of the vaccinations are meant to be calculated and included in the numerator - but everything else looks good!
